### PR TITLE
transfer-contract: Update codebase to canonical 0.6

### LINKED ISSUE #305: Test transfer contract with the new stack

### PR DESCRIPTION
**Pull request by vlopes11**
Update transfer contract to use the stack of canonical 0.6, including
the newly released dusk-hamt.

Resolves #305

### CODE REVIEW

File: contracts/transfer/src/wasm/internal.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/internal.rs:211]: Is really needed?

Comment (CPerezz) [reply contracts/transfer/src/wasm/internal.rs:211]: And if the interface is, why is not being called?

**Review by CPerezz — COMMENTED** [review]
A few questions/nits.

Looks fine overall. Once clarified, I'll approve!

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:48]: This is a replacement for the spurious functions

https://github.com/dusk-network/rusk/blob/vlopes11/transfer-contract/contracts/transfer/src/wasm/bridge.rs#L16

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:97]: We also need to keep a set of all existing roots as a part of the state. This function exists to enforce that rule

https://github.com/dusk-network/rusk/blob/vlopes11/transfer-contract/contracts/transfer/src/transfer.rs#L33

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:78]: Same as above; this is the approach for spurious functions aren't applicable for the transfer contract context

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:82]: Same as above; this is the approach for spurious functions aren't applicable for the transfer contract context

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:86]: Same as above; this is the approach for spurious functions aren't applicable for the transfer contract context

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:211]: The proof verification method constantly changes, so that is a way to encapsulate all these changes into a single function

Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:211]: But shouldn't be commented out, that is a mistake. Will enable again

File: contracts/transfer/Makefile
Comment (ZER0) [reply contracts/transfer/Makefile:9]: It's ok, since we're hotfixing master anyway.

File: contracts/transfer/src/transfer.rs
Comment (ZER0) [reply contracts/transfer/src/transfer.rs:48]: What you mean "replacement"? It's been a while you can actually use "spurious" function. But I believe the actual spurious function needs the view key no?

File: contracts/transfer/src/transfer.rs
Comment (ZER0) [reply contracts/transfer/src/transfer.rs:78]: I'm not sure I'm following: why aren't applicable for the transfer contract context…?

File: contracts/transfer/src/wasm/internal.rs
Comment (ZER0) [inline contracts/transfer/src/wasm/internal.rs:209]: Shouldn't we fix that now?

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM, with the typo in the commit message fixed.
There are, however, few things not totally clear to me. Are the actual "spurious" function implemented? Because I didn't see it but maybe are not part of this PR? The same for the tests related to them.

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs]: Fixed

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs]: Yep, was a nit

**Comment by CPerezz** [comment]
@vlopes11 the commit message of b6a269f references a PR instead of the issue that you wanted to reference.

File: contracts/transfer/Makefile
Comment (vlopes11) [reply contracts/transfer/Makefile]: Fixed

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:49]: As discussed, spurious functions are getters of the state

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:63]: Ditto

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs]: Notes mapping was removed

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs]: Ditto

File: contracts/transfer/src/transfer/tree.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer/tree.rs]: I want to implement custom functions to this type. It is better we have it over an entirely new type and encapsulate the tree as backend

File: contracts/transfer/src/transfer/tree.rs
Comment (CPerezz) [inline contracts/transfer/src/transfer/tree.rs:187]: You can already address this here.

File: contracts/transfer/src/wasm/internal.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/internal.rs:24]: Isn't the host function avaliable in this abi version?

File: contracts/transfer/src/wasm/internal.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/internal.rs:42]: This is not contract's responsability but VM's one. We should save this info there since on the contract we can never change it if we want to in the future.

File: rust-toolchain
Comment (CPerezz) [inline rust-toolchain:1]: Set `nightly-2021-06-06`

**Review by CPerezz — CHANGES_REQUESTED** [review]
Some changes and questions.

File: circuits/transfer/src/send_to_contract_obfuscated.rs
Comment (vlopes11) [reply circuits/transfer/src/send_to_contract_obfuscated.rs:261]: This proposal is only about code styling - and in lining usually degrade the readability

File: contracts/transfer/src/transfer.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer.rs:103]: @ZER0 do we have an issue for the change you are performing?

File: contracts/transfer/src/transfer/tree.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer/tree.rs:187]: Not clear

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:24]: Probably yes. What is the current ABI call?

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:42]: AFAIK, we don't have such function available in the ABI

File: contracts/transfer/src/transfer.rs
Comment (ZER0) [reply contracts/transfer/src/transfer.rs:103]: I do: https://github.com/dusk-network/cargo-bake/issues/1

File: circuits/transfer/src/withdraw_from_transparent.rs
Comment (vlopes11) [reply circuits/transfer/src/withdraw_from_transparent.rs:11]: Good catch, fixed

File: contracts/transfer/Cargo.toml
Comment (vlopes11) [reply contracts/transfer/Cargo.toml:32]: Fixed

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:24]: Good catch, fixed

File: rust-toolchain
Comment (vlopes11) [reply rust-toolchain]: Fixed

**Review by ZER0 — CHANGES_REQUESTED** [review]
There are some nits that should be addressed, but as mentioned that would be done in a follow up PR.

So I kept here only the major concerns that I think we should discuss / address before merging this, such as the gas and the feature flags.

File: contracts/transfer/tests/wrapper/alice/src/lib.rs
Comment (vlopes11) [reply contracts/transfer/tests/wrapper/alice/src/lib.rs:50]: This is a test contract. It is testing the capability of performing a single call - the debug is useful in case of a panic that could or could not have called the contract query

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:216]: The `assert_payable` function? Its part of the core functionality of the contract

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:34]: This part will change and, as discussed, is waiting for the availability of the `gas_left` ABI call

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:32]: Ditto; This part will change and, as discussed, is waiting for the availability of the gas_left ABI call

File: contracts/transfer/src/wasm/callable.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/callable.rs:272]: How is that? We need to evaluate the minimum gas price, it is one of the documented execution steps

File: contracts/transfer/Cargo.toml
Comment (vlopes11) [reply contracts/transfer/Cargo.toml:33]: I need to import the types in wasm32 to perform the execute calls. This way, I need to use the contract as dependency without declaring the `q` and `t` `no_mangle` functions

File: contracts/transfer/src/wasm.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm.rs:7]: Ditto

File: contracts/transfer/tests/wrapper/alice/Cargo.toml
Comment (vlopes11) [reply contracts/transfer/tests/wrapper/alice/Cargo.toml:17]: Ditto

File: circuits/transfer/src/send_to_contract_obfuscated.rs
Comment (vlopes11) [reply circuits/transfer/src/send_to_contract_obfuscated.rs:261]: N/A to this PR

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:44]: As discussed via call, its applicable and we just need to append documents and mention this is always in nano-dusk

File: contracts/transfer/Cargo.toml
Comment (vlopes11) [reply contracts/transfer/Cargo.toml:33]: As discussed, we will land as is and in the future this will probably be handled by the ABI

File: contracts/transfer/src/transfer/tree.rs
Comment (vlopes11) [reply contracts/transfer/src/transfer/tree.rs:187]: Good catch, fixed

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:44]: Fixed

File: contracts/transfer/src/wasm/internal.rs
Comment (vlopes11) [reply contracts/transfer/src/wasm/internal.rs:216]: Fixed

File: circuits/transfer/src/gadgets.rs
Comment (CPerezz) [inline circuits/transfer/src/gadgets.rs:49]: Since this is more generic, we could move that to `plonk_gadgets`. Is fine to merge now if you want. But file an issue to migrate it at some point.

File: contracts/transfer/src/transfer.rs
Comment (CPerezz) [inline contracts/transfer/src/transfer.rs:28]: Maybe file an issue to not forget about this?

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:44]: You don't care wether is `Some(key) `or `None`?

File: contracts/transfer/src/wasm/callable.rs
Comment (CPerezz) [inline contracts/transfer/src/wasm/callable.rs:162]: If is not needed, remove it.

File: contracts/transfer/src/wasm/internal.rs
Comment (CPerezz) [reply contracts/transfer/src/wasm/internal.rs:29]: Maybe add an issue for it referring to this part would be nice.

File: rusk-profile/src/lib.rs
Comment (CPerezz) [inline rusk-profile/src/lib.rs:16]: Is this change correct? Is it related to the latest PR @ZER0 did I guess right?

**Review by CPerezz — APPROVED** [review]
A couple questions and nits but overall looks good after addressing them.

Still concerned about the `no-bridge` feature.. But we will figure out something. This needs to land so that we can develop the other contracts.

**Review by ZER0 — APPROVED** [review]

**Pull request merged**
